crates/bevy_ecs/src/component.rs

```diff
@@ -952,6 +1000,278 @@ impl Components {
+        unsafe {
+            self.get_recursive_required_components(required, &mut recursive_requires, 1);
+        };
+
+        // Register the new required components for the requiree.
+        // SAFETY: The caller ensures that the `requiree` is valid.
+        //         The recursive required components are valid, since they were just found.
+        unsafe {
+            self.register_required_components_internal(requiree, &recursive_requires);
+        }
+
+        // Propagate the new required components up the chain to all components that require the requiree.
+        if let Some(required_by) = self.get_required_by(requiree).cloned() {
+            for &required_by_id in required_by.iter() {
+                // SAFETY: The component is in the list of required components, so it must exist already.
+                let required_components = unsafe {
+                    self.get_required_components_mut(required_by_id)
+                        .debug_checked_unwrap()
+                };
+
+                // Register the original required component for the requiree.
+                // The inheritance depth is `1` since this is a component required by the original requiree.
+                required_components.register_by_id(required, constructor, 1);
+
+                for (component_id, component) in recursive_requires.iter() {
+                    // Register the required component.
+                    // The inheritance depth is increased by `1` since this is a component required by the original required component.
+                    // SAFETY: Component ID and constructor match the ones on the original requiree.
+                    //         The original requiree is responsible for making sure the registration is safe.
+                    unsafe {
+                        required_components.register_dynamic(
+                            *component_id,
+                            component.constructor.clone(),
+                            component.inheritance_depth + 1,
+                        );
+                    };
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Recursively finds all required components for a given component,
+    /// adding them to the given `recursive_requires` list.
+    ///
+    /// # Safety
+    ///
+    /// The given `id` must be valid.
+    unsafe fn get_recursive_required_components(
+        &self,
+        id: ComponentId,
+        recursive_requires: &mut Vec<(ComponentId, RequiredComponent)>,
+        depth: u16,
+    ) {
+        // Get directly required components at this level of the inheritance tree.
+        // SAFETY: The caller ensures that the `id` is valid.
+        let component_info = unsafe { self.get_info(id).debug_checked_unwrap() };
+        let required_components = component_info
+            .required_components()
+            .0
+            .iter()
+            .filter(|(_, component)| component.inheritance_depth == 0);
```

Comment:
I found a bug here with how the depths are computed. This `required_components` variable didn't have just the direct requirements, but *all* requirements in the tree, which made the recursion redundant and the inheritance depths wrong.

I fixed it by adding this filter, and added a regression test. I also realized that we could even do this non-recursively, and just iterate through depths and the components at those depths until no components are left. Not sure which approach is better however.

Reply:
Wait I think I'm being dumb lmao, we don't need any recursion or complicated logic here, the required components should already be correct, we just need to add 1 to the depth of all of them when registering them for the requiree here. Fixing in a moment

Reply:
Okay I simplified the logic now, removing the recursion and unnecessary work. Tests still pass, so I think it should be good.